lib/concurrent/synchronization/volatile.rb

```diff
@@ -0,0 +1,35 @@
+module Concurrent
+  module Synchronization
+
+    # Volatile adds the attr_volatile class method when included.
+    #
+    # @example
+    #   class Foo
+    #     include Concurrent::Synchronization::Volatile
+    #
+    #     attr_volatile :bar
+    #
+    #     def initialize
+    #       self.bar = 1
+    #     end
+    #   end
+    #
+    #  foo = Foo.new
+    #  foo.bar
+    #  => 1
+    #  foo.bar = 2
+    #  => 2
+
+    Volatile = case
+    when Concurrent.on_cruby?
+      MriAttrVolatile
+    when Concurrent.on_jruby?
+      JRubyAttrVolatile
+    when Concurrent.on_rbx?
+      RbxAttrVolatile
+    else
+      warn 'Possibly unsupported Ruby implementation'
+      MriAttrVolatile
+    end
```

Comment:
not sure exactly what you are suggesting: I see that in Object there is `private_constant :ObjectImplementation`, but in `volatile.rb` the `Volatile` constant must stay public since the it the module you include, which points to the correct module implementation.

or do you mean just adding
```ruby
    # @!visibility private
    # @!macro internal_implementation_note
```
to the actual module implementation, like `module JRubyAttrVolatile` ?